File: clojure/grade-school/src/grade_school.clj
(ns grade-school)

(defn add [db name grade]
  (assoc db grade (conj (vec (db grade)) name)))

(defn grade [db grade]
  (if-let [names (db grade)] names []))

(defn sorted [db]
  (let [sort-val (fn [db key] (assoc db key (vec (sort (db key)))))]
    (into (sorted-map) (reduce sort-val db (keys db)))))

File: javascript/grade-school/grade-school.js
module.exports = function School() {
  var students = {};

  function add(name, grade) {
    if(students[grade]) {
      students[grade].push(name);
    } else {
      students[grade] = [name];
    }
  }

  function grade(grade) {
      return students[grade] ? clone(students[grade]).sort() : [];
  }

  function roster() {
    var sortedGrade = Object.keys(students).sort();
    return sortedGrade.reduce(function (sorted, grade) {
      sorted[grade] = clone(students[grade]).sort();
      return sorted;
    }, {});
  }

  return {
    add: add,
    grade: grade,
    roster: roster
  };
};

function clone(array) {
  return array.slice();
}
